File: leetcode/230.二叉搜索树中第K小的元素/230.二叉搜索树中第K小的元素.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    """
    根据BST的特性，左小右大，子树依然，使用中序遍历找到第k个数
    """

    def __init__(self):
        self.res = 0
        # 排名
        self.rank = 0

    def kthSmallest(self, root: TreeNode, k: int) -> int:
        self.traverse(root, k)
        return self.res

    def traverse(self, root, k):
        if root is None:
            return root
        self.traverse(root, k)
        self.rank += 1
        if k == self.rank:
            self.res = root.val
        self.traverse(root, k)

File: leetcode/230.二叉搜索树中第K小的元素/230.二叉搜索树中第K小的元素.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func kthSmallest(root *TreeNode, k int) int {
	res := 0
	rank := 0

	var reverse func(root *TreeNode, k int)

	reverse = func(root *TreeNode, k int) {
		if root == nil {
			return
		}
		reverse(root.Left, k)
		rank += 1
		if rank == k {
			res = root.Val
		}
		reverse(root.Right, k)
	}
	reverse(root, k)
	return res
}
